### PR TITLE
Drs4 time corr

### PR DESCRIPTION
**Pull request by pawel21**
Hi,

I have created code to DRS4 time correction.
We have two class `TimeCorrectionCalculate` to create file with coefficients for time correction curve
of chip DRS4 and `PulseTimeCorrection` to make time correction on pulse time from extractor. 
I added notebook `time_calibration.ipynb` which show how to create h5py file and how to apply correction to camera calibration run and cosmic events.

@FrancaCassol Please look on my code and tell what do you think? 
I think, we should ,,connect" time correction with flatfield calibration 

What is the status of gain selection? Because, now code is designed for waveform with both gains.

### CODE REVIEW

**Comment by FrancaCassol** [comment]
Hi @pawel21,

I can not test the code before this evening, some questions for the moment:
- could you please describe the method and algorithms you use for deriving and applying the corrections?
- In the notebook I see you modify not only the width but also the mean value of the time, why?
- is your correction dependent from the charge extractor you use?
- you speak of h5py, but I guess it is a normal hdf5 file, isn't it? (in this case better you speak of hdf5)

**Comment by pawel21** [comment]
> Hi @pawel21,
> 
> I can not test the code before this evening, some questions for the moment:
No problem.
> 
>     * could you please describe the method and algorithms you use for deriving and applying the corrections?
,,DRS4 chips exhibit an delay of typically 1 ns (up to 4 ns), depending on the absolute location of the signal pulse in the domino ring. The differences can be calibrated with the use of calibration pulses. The computed mean arrival time of a pulse in a channel is a rather complicated function of the time sample position in the ring and differs from one DRS chip to another. For each channel we expand this function into a Fourier series to obtain the correction function." [from https://arxiv.org/abs/1305.1007]  To computed mean arrival time of a pulse I am using `LocalPeakWindowSum` (`window_width` and `window_shift` user can set up), then I fit data using Fourier series, I obtain coefficients, which I save to hdf5 file. Then, I am using "correction curve" (which coefficients from previous step) to correct arrival time (pulse time) from extractor in each pixel. For single event time resolution is decreased from 1.6 ns to 0.4 ns.

> 
>     * In the notebook I see you modify not only the width but also the mean value of the time, why?
>
The idea of time correction is the following: we got arrival time from calibration pulse (`charge, pulse_time = extractor(ev.r1.tel[1].waveform`) and then using "correction curve" in class `PulseTimeCorrection` we got arrival time after correction `pulse_corr_array = PulseTimeCorrection.get_corr_pulse(ev, pulse_time)`. So, my code modify `pulse_time`
 
>     * is your correction dependent from the charge extractor you use?
Yes, this is not big differences. User can set up parameters for `LocalPeakWindowSum` (I added example in notebook"). It is best to use the same parameters for calibration and for processing data.
> 
>     * you speak of h5py, but I guess it is a normal hdf5 file, isn't it? (in this case better you speak of hdf5)
Yes, I use normal hdf5 file. Thank you.

**Comment by FrancaCassol** [comment]
Hi @pawel21,

thanks for your explanation. Could you please add the link to the MAGIC document to the doc of the of the classes?

Also, I wonder, is this correction compatible with the sampling time correction of Yuto? Do you know if the correction proposed by Yuto is applied in Magic?

**Comment by pawel21** [comment]
> Hi @pawel21,
> 
> thanks for your explanation. Could you please add the link to the MAGIC document to the doc of the of the classes?

Yes, sure

> 
> Also, I wonder, is this correction compatible with the sampling time correction of Yuto? Do you know if the correction proposed by Yuto is applied in Magic?

Yes, should be compatible. My method will be applied to time, and "sampling time correction" meted will be applied to the value of the reproduced signal, then they should not bite. 
Sampling time correction of Yuko is used to improve the extracted signal. This method is not used in MAGIC.

File: lstchain/calib/camera/time_correction_calculate.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    minimum_charge = Float(200,
+                           help='Cut on charge. Default 200 ADC'
+                          ).tag(config=True)
+
+    r1_sample_start = Int(default_value=2,
+                          help='Start sample for r1 waveform',
+                          allow_none=True).tag(config=True)
+
+
+    r1_sample_end = Int(default_value=38,
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/calib/camera/time_correction_calculate.py]: Yes, You are right. Thank you

File: lstchain/calib/camera/time_correction_calculate.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+    def calibrate_pulse_time(self, event):
+        """
+        Fill bins using time pulse from LocalPeakWindowSum.
+        Parameters
+        ----------
+        event : `ctapipe` event-container
+        """
+        if event.r0.tel[self.tel_id].trigger_type == 1:
+            for nr_module in prange(0, n_modules):
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/calib/camera/time_correction_calculate.py]: What is the difference between trigger and r1 trigger? But, I agree use r1 trigger will be more convenient

File: lstchain/calib/camera/time_correction_calculate.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                self.first_cap_array[nr_module, :, :] = self.get_first_capacitor(event, nr_module)
+
+            pixel_ids = event.lst.tel[self.tel_id].svc.pixel_ids
+            charge, pulse_time = self.extractor(event.r1.tel[self.tel_id].waveform[:, :, self.r1_sample_start:self.r1_sample_end])
+            self.calib_pulse_time_jit(charge,
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/calib/camera/time_correction_calculate.py]: I removed. Now, my code use r1 samples. Thank you

File: lstchain/calib/camera/time_correction_calculate.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+
+__all__ = ['TimeCorrection']
+
+high_gain = 0
+low_gain = 1
+
+n_gain = 2
+n_channel = 7
+n_modules = 265
```
[the remark this answers is not in the document]
Comment (pawel21) [reply lstchain/calib/camera/time_correction_calculate.py]: This wasn't compatible. I have modified code to make correction only for working modules (`        n_modules_from_event = event.lst.tel[self.tel_id].svc.num_modules` line 69 in `pulse_time_correction.py`)

File: lstchain/calib/camera/time_correction_calculate.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrancaCassol) [reply lstchain/calib/camera/time_correction_calculate.py]: Hi @pawel21,
there is not difference, at least for the moment. But, after the r0 calibration, I think it is better to work only with r1 elements

**Comment by rlopezcoto** [comment]
Thanks @pawel21 for the code and @FrancaCassol for the review. This PR is very necessary to improve signal extraction.

**Pull request merged**